radix-transactions/src/manifest/static_resource_movements/types.rs

```diff
@@ -0,0 +1,545 @@
+            }
+            // If the lower bound is non-zero and the upper bound is defined then we might need to
+            // update both the upper and lower bound.
+            (
+                ref mut lower_bound @ LowerFungibleBound::NonZero,
+                UpperFungibleBound::Amount(ref mut existing_upper_bound),
+            ) => {
+                *lower_bound = LowerFungibleBound::Amount(new_lower_bound);
+                *existing_upper_bound = new_lower_bound.max(*existing_upper_bound);
+            }
+            // If a lower bound is defined and the upper bound is not defined then the new lower
+            // bound would be the maximum of the existing lower bound and the new lower bound.
+            (
+                ref mut lower_bound @ LowerFungibleBound::Amount(existing_lower_bound),
+                UpperFungibleBound::Unbounded,
+            ) => {
+                *lower_bound = LowerFungibleBound::Amount(existing_lower_bound.max(new_lower_bound))
+            }
+            // If both a lower and upper bound are numerically defined then we do the following:
+            // - Set the lower bound to the max of the existing lower bound and the new lower bound.
+            // - Set the upper bound to tbe the maximum of the new lower bound and the existing
+            //   upper bound.
+            // This is done to move both the lower and upper bounds in cases where they need to be
+            // moved.
+            (
+                LowerFungibleBound::Amount(ref mut existing_lower_bound),
+                UpperFungibleBound::Amount(ref mut existing_upper_bound),
+            ) => {
+                *existing_lower_bound = (*existing_lower_bound).max(new_lower_bound);
+                *existing_upper_bound = (*existing_upper_bound).max(*existing_lower_bound);
+            }
+        }
+    }
+
+    pub(super) fn decrease_both_bounds(&mut self, by: Decimal) -> Option<()> {
+        match (self.lower, self.upper) {
+            // The upper bound is being reduced by some amount. We first start by reducing the upper
+            // bound and then check to ensure that it is not zero. If it is equal to zero then the
+            // bounds will be changed to an exact of zero.
+            (LowerFungibleBound::NonZero, UpperFungibleBound::Amount(ref mut upper_bound)) => {
+                // Reduce the upper bound.
+                *upper_bound = upper_bound.checked_sub(by)?.max(Decimal::ZERO);
+
+                // Check if the upper bound is now zero. If it is, then we switch to being an exact
+                // zero.
+                if *upper_bound <= Decimal::ZERO {
+                    *self = Self::new_exact(Decimal::ZERO);
+                }
+            }
+            // The lower bound is not zero and the upper bound is unbounded. Can't do anything at
+            // all here.
+            (LowerFungibleBound::NonZero, UpperFungibleBound::Unbounded) => {}
```

Comment:
(Addressed in https://github.com/radixdlt/radixdlt-scrypto/pull/1931 )

Reply:
Why would we set the lower bound to zero here? 

> Honestly, I suggest LowerFungibleBound::NonZero gets deleted and we have a LowerFungibleBound::non_zero() method which maps to LowerFungibleBound::Amount(Decimal::of_subunits(1)) - it would make handling all these cases easier :)

I personally really do not like having special cased magic numbers and find the `NonZero` enum variant to be more expressive.

Reply:
Address in #1931 pairing with you :+1: